File: app/controllers/ExamSectionController.java
/*
 * Copyright (c) 2018 The members of the EXAM Consortium (https://confluence.csc.fi/display/EXAM/Konsortio-organisaatio)
 *
 * Licensed under the EUPL, Version 1.1 or - as soon they will be approved by the European Commission - subsequent
 * versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl/licence-eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed
 * on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package controllers;

import be.objectify.deadbolt.java.actions.Group;
import be.objectify.deadbolt.java.actions.Restrict;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import controllers.base.BaseController;
import controllers.base.SectionQuestionHandler;
import impl.ExamUpdaterImpl;
import io.ebean.DB;
import io.ebean.ExpressionList;
import io.ebean.text.PathProperties;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.TreeSet;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;
import javax.inject.Inject;
import models.Exam;
import models.Role;
import models.User;
import models.questions.MultipleChoiceOption;
import models.questions.MultipleChoiceOption.ClaimChoiceOptionType;
import models.questions.Question;
import models.sections.ExamSection;
import models.sections.ExamSectionQuestion;
import models.sections.ExamSectionQuestionOption;
import org.joda.time.DateTime;
import play.data.DynamicForm;
import play.db.ebean.Transactional;
import play.libs.Json;
import play.mvc.Http;
import play.mvc.Result;
import play.mvc.With;
import sanitizers.Attrs;
import sanitizers.SanitizingHelper;
import sanitizers.SectionQuestionSanitizer;
import security.Authenticated;

public class ExamSectionController extends BaseController implements SectionQuestionHandler {

    @Inject
    private ExamUpdaterImpl examUpdater;

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result insertSection(Long id, Http.Request request) {
        Optional<Exam> oe = DB
            .find(Exam.class)
            .fetch("examOwners")
            .fetch("examSections")
            .where()
            .idEq(id)
            .findOneOrEmpty();
        if (oe.isEmpty()) {
            return notFound("sitnet_error_not_found");
        }
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        Exam exam = oe.get();
        // Not allowed to add a section if optional sections exist and there are upcoming reservations
        boolean optionalSectionsExist = exam.getExamSections().stream().anyMatch(ExamSection::isOptional);
        if (optionalSectionsExist && !examUpdater.isAllowedToUpdate(exam, user)) {
            return forbidden("sitnet_error_future_reservations_exist");
        }
        if (exam.isOwnedOrCreatedBy(user) || user.hasRole(Role.Name.ADMIN)) {
            ExamSection section = new ExamSection();
            section.setLotteryItemCount(1);
            section.setExam(exam);
            section.setSectionQuestions(Collections.emptySet());
            section.setSequenceNumber(exam.getExamSections().size());
            section.setExpanded(true);
            section.setOptional(false);
            section.setCreatorWithDate(user);
            section.save();
            return ok(section, PathProperties.parse("(*, examMaterials(*), sectionQuestions(*))"));
        } else {
            return forbidden("sitnet_error_access_forbidden");
        }
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result removeSection(Long eid, Long sid, Http.Request request) {
        Optional<Exam> oe = DB
            .find(Exam.class)
            .fetch("examOwners")
            .fetch("examSections")
            .where()
            .idEq(eid)
            .findOneOrEmpty();
        ExamSection section = DB.find(ExamSection.class, sid);
        if (oe.isEmpty() || section == null) {
            return notFound("sitnet_error_not_found");
        }
        Exam exam = oe.get();
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        // Not allowed to remove a section if optional sections exist and there are upcoming reservations
        boolean optionalSectionsExist = exam.getExamSections().stream().anyMatch(ExamSection::isOptional);
        if (optionalSectionsExist && !examUpdater.isAllowedToUpdate(exam, user)) {
            return forbidden("sitnet_error_future_reservations_exist");
        }
        if (exam.isOwnedOrCreatedBy(user) || user.hasRole(Role.Name.ADMIN)) {
            exam.getExamSections().remove(section);
            // Decrease sequences for the entries above the inserted one
            int seq = section.getSequenceNumber();
            for (ExamSection es : exam.getExamSections()) {
                int num = es.getSequenceNumber();
                if (num >= seq) {
                    es.setSequenceNumber(num - 1);
                    es.update();
                }
            }
            section.delete();
            return ok();
        } else {
            return forbidden("sitnet_error_access_forbidden");
        }
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result updateSection(Long eid, Long sid, Http.Request request) {
        Optional<Exam> oe = DB
            .find(Exam.class)
            .fetch("examOwners")
            .fetch("examSections")
            .where()
            .idEq(eid)
            .findOneOrEmpty();
        ExamSection section = DB.find(ExamSection.class, sid);
        if (oe.isEmpty() || section == null) {
            return notFound("sitnet_error_not_found");
        }
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        if (!oe.get().isOwnedOrCreatedBy(user) && !user.hasRole(Role.Name.ADMIN)) {
            return unauthorized("sitnet_error_access_forbidden");
        }

        ExamSection form = formFactory
            .form(ExamSection.class)
            .bindFromRequest(
                request,
                "id",
                "name",
                "expanded",
                "lotteryOn",
                "lotteryItemCount",
                "description",
                "optional"
            )
            .get();

        section.setName(form.getName());
        section.setExpanded(form.isExpanded());
        section.setLotteryOn(form.isLotteryOn());
        section.setLotteryItemCount(Math.max(1, form.getLotteryItemCount()));
        section.setDescription(form.getDescription());
        // Disallow changing optionality if future reservations exist
        if (section.isOptional() != form.isOptional() && !examUpdater.isAllowedToUpdate(section.getExam(), user)) {
            return badRequest("sitnet_error_future_reservations_exist");
        }

        section.setOptional(form.isOptional());

        section.update();

        return ok(section);
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result reorderSections(Long eid, Http.Request request) {
        DynamicForm df = formFactory.form().bindFromRequest(request);
        int from = Integer.parseInt(df.get("from"));
        int to = Integer.parseInt(df.get("to"));
        return checkBounds(from, to)
            .orElseGet(() -> {
                Exam exam = DB.find(Exam.class).fetch("examSections").where().idEq(eid).findOne();
                if (exam == null) {
                    return notFound("sitnet_error_exam_not_found");
                }
                User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
                if (exam.isOwnedOrCreatedBy(user) || user.hasRole(Role.Name.ADMIN)) {
                    // Reorder by sequenceNumber (TreeSet orders the collection based on it)
                    List<ExamSection> sections = new ArrayList<>(new TreeSet<>(exam.getExamSections()));
                    ExamSection prev = sections.get(from);
                    boolean removed = sections.remove(prev);
                    if (removed) {
                        sections.add(to, prev);
                        for (int i = 0; i < sections.size(); ++i) {
                            ExamSection section = sections.get(i);
                            section.setSequenceNumber(i);
                            section.update();
                        }
                    }
                    return ok();
                }
                return forbidden("sitnet_error_access_forbidden");
            });
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result reorderSectionQuestions(Long eid, Long sid, Http.Request request) {
        DynamicForm df = formFactory.form().bindFromRequest(request);
        int from = Integer.parseInt(df.get("from"));
        int to = Integer.parseInt(df.get("to"));
        return checkBounds(from, to)
            .orElseGet(() -> {
                Exam exam = DB.find(Exam.class, eid);
                if (exam == null) {
                    return notFound("sitnet_error_exam_not_found");
                }
                User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
                if (exam.isOwnedOrCreatedBy(user) || user.hasRole(Role.Name.ADMIN)) {
                    ExamSection section = DB.find(ExamSection.class, sid);
                    if (section == null) {
                        return notFound("section not found");
                    }
                    // Reorder by sequenceNumber (TreeSet orders the collection based on it)
                    List<ExamSectionQuestion> questions = new ArrayList<>(new TreeSet<>(section.getSectionQuestions()));
                    ExamSectionQuestion prev = questions.get(from);
                    boolean removed = questions.remove(prev);
                    if (removed) {
                        questions.add(to, prev);
                        for (int i = 0; i < questions.size(); ++i) {
                            ExamSectionQuestion question = questions.get(i);
                            question.setSequenceNumber(i);
                            question.update();
                        }
                    }
                    return ok();
                }
                return forbidden("sitnet_error_access_forbidden");
            });
    }

    private void updateExamQuestion(ExamSectionQuestion sectionQuestion, JsonNode body, Http.Request request) {
        sectionQuestion.setMaxScore(round(SanitizingHelper.parse("maxScore", body, Double.class).orElse(null)));
        sectionQuestion.setAnswerInstructions(request.attrs().getOptional(Attrs.ANSWER_INSTRUCTIONS).orElse(null));
        sectionQuestion.setEvaluationCriteria(request.attrs().getOptional(Attrs.EVALUATION_CRITERIA).orElse(null));
        sectionQuestion.setEvaluationType(
            SanitizingHelper.parseEnum("evaluationType", body, Question.EvaluationType.class).orElse(null)
        );
        sectionQuestion.setExpectedWordCount(
            SanitizingHelper.parse("expectedWordCount", body, Integer.class).orElse(null)
        );
    }

    private Optional<Result> insertQuestion(Exam exam, ExamSection section, Question question, User user, Integer seq) {
        ExamSectionQuestion sectionQuestion = new ExamSectionQuestion();
        sectionQuestion.setExamSection(section);
        sectionQuestion.setQuestion(question);
        // Assert that the sequence number provided is within limits
        int sequence = Math.min(Math.max(0, seq), section.getSectionQuestions().size());
        updateSequences(section.getSectionQuestions(), sequence);
        sectionQuestion.setSequenceNumber(sequence);
        if (section.getSectionQuestions().contains(sectionQuestion) || section.hasQuestion(question)) {
            return Optional.of(badRequest("sitnet_question_already_in_section"));
        }
        if (question.getType().equals(Question.Type.EssayQuestion)) {
            // disable auto evaluation for this exam
            if (exam.getAutoEvaluationConfig() != null) {
                exam.getAutoEvaluationConfig().delete();
            }
        }

        DB.updateAll(section.getSectionQuestions());

        // Insert new section question
        sectionQuestion.setCreator(user);
        sectionQuestion.setCreated(DateTime.now());
        sectionQuestion.setExamSection(section);

        updateExamQuestion(sectionQuestion, question);

        section.getSectionQuestions().add(sectionQuestion);

        section.setModifierWithDate(user);
        section.save();
        section.setSectionQuestions(new TreeSet<>(section.getSectionQuestions()));
        return Optional.empty();
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result insertQuestion(Long eid, Long sid, Long qid, Http.Request request) {
        Exam exam = DB.find(Exam.class, eid);
        ExamSection section = DB.find(ExamSection.class, sid);
        Question question = DB.find(Question.class, qid);
        if (exam == null || section == null || question == null) {
            return notFound();
        }
        if (exam.getAutoEvaluationConfig() != null && question.getType() == Question.Type.EssayQuestion) {
            return forbidden("sitnet_error_autoevaluation_essay_question");
        }
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        if (!exam.isOwnedOrCreatedBy(user) && !user.hasRole(Role.Name.ADMIN)) {
            return forbidden("sitnet_error_access_forbidden");
        }
        Integer seq = request.body().asJson().get("sequenceNumber").asInt();
        return insertQuestion(exam, section, question, user, seq).orElse(ok(section));
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    @Transactional
    public Result insertMultipleQuestions(Long eid, Long sid, Http.Request request) {
        Exam exam = DB.find(Exam.class, eid);
        ExamSection section = DB.find(ExamSection.class, sid);
        if (exam == null || section == null) {
            return notFound();
        }
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        if (!exam.isOwnedOrCreatedBy(user) && !user.hasRole(Role.Name.ADMIN)) {
            return forbidden("sitnet_error_access_forbidden");
        }
        int sequence = request.body().asJson().get("sequenceNumber").asInt();
        String questions = request.body().asJson().get("questions").asText();
        for (String s : questions.split(",")) {
            Question question = DB.find(Question.class, Long.parseLong(s));
            if (question == null) {
                continue;
            }
            if (exam.getAutoEvaluationConfig() != null && question.getType() == Question.Type.EssayQuestion) {
                return forbidden("sitnet_error_autoevaluation_essay_question");
            }
            Optional<Result> result = insertQuestion(exam, section, question, user, sequence);
            if (result.isPresent()) {
                return result.get();
            }
            ++sequence;
        }
        return ok(section);
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result removeQuestion(Long eid, Long sid, Long qid, Http.Request request) {
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        ExamSectionQuestion sectionQuestion = DB
            .find(ExamSectionQuestion.class)
            .fetch("examSection.exam.examOwners")
            .fetch("question")
            .where()
            .eq("examSection.exam.id", eid)
            .eq("examSection.id", sid)
            .eq("question.id", qid)
            .findOne();
        if (sectionQuestion == null) {
            return notFound("sitnet_error_not_found");
        }
        ExamSection section = sectionQuestion.getExamSection();
        Exam exam = section.getExam();
        if (!exam.isOwnedOrCreatedBy(user) && !user.hasRole(Role.Name.ADMIN)) {
            return forbidden("sitnet_error_access_forbidden");
        }
        section.getSectionQuestions().remove(sectionQuestion);

        // Decrease sequences for the entries above the inserted one
        int seq = sectionQuestion.getSequenceNumber();
        for (ExamSectionQuestion esq : section.getSectionQuestions()) {
            int num = esq.getSequenceNumber();
            if (num >= seq) {
                esq.setSequenceNumber(num - 1);
                esq.update();
            }
        }
        // Update lottery item count if needed
        if (section.isLotteryOn() && section.getLotteryItemCount() > section.getSectionQuestions().size()) {
            section.setLotteryItemCount(section.getSectionQuestions().size());
        }
        sectionQuestion.delete();
        return ok(section);
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result clearQuestions(Long eid, Long sid, Http.Request request) {
        ExamSection section = DB
            .find(ExamSection.class)
            .fetch("exam.creator")
            .fetch("exam.examOwners")
            .fetch("exam.parent.examOwners")
            .where()
            .idEq(sid)
            .eq("exam.id", eid)
            .findOne();
        if (section == null) {
            return notFound("sitnet_error_not_found");
        }
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        if (section.getExam().isOwnedOrCreatedBy(user) || user.hasRole(Role.Name.ADMIN)) {
            section
                .getSectionQuestions()
                .forEach(sq -> {
                    sq
                        .getQuestion()
                        .getChildren()
                        .forEach(c -> {
                            c.setParent(null);
                            c.update();
                        });
                    sq.delete();
                });
            section.getSectionQuestions().clear();
            section.setLotteryOn(false);
            section.update();
            return ok(section);
        } else {
            return forbidden("sitnet_error_access_forbidden");
        }
    }

    private void createOptionBasedOnExamQuestion(Question question, ExamSectionQuestion esq, User user, JsonNode node) {
        MultipleChoiceOption option = new MultipleChoiceOption();
        JsonNode baseOptionNode = node.get("option");
        option.setOption(SanitizingHelper.parse("option", baseOptionNode, String.class).orElse(null));
        option.setDefaultScore(round(SanitizingHelper.parse("score", node, Double.class).orElse(null)));
        Boolean correctOption = SanitizingHelper.parse("correctOption", baseOptionNode, Boolean.class, false);
        option.setCorrectOption(correctOption);
        saveOption(option, question, user);
        propagateOptionCreationToExamQuestions(question, esq, option);
    }

    private void processExamQuestionOptions(Question question, ExamSectionQuestion esq, ArrayNode node, User user) { // esq.options
        Set<Long> persistedIds = question
            .getOptions()
            .stream()
            .map(MultipleChoiceOption::getId)
            .collect(Collectors.toSet());
        Set<Long> providedIds = StreamSupport
            .stream(node.spliterator(), false)
            .map(n -> n.get("option"))
            .filter(n -> SanitizingHelper.parse("id", n, Long.class).isPresent())
            .map(n -> SanitizingHelper.parse("id", n, Long.class).get())
            .collect(Collectors.toSet());
        // Updates
        StreamSupport
            .stream(node.spliterator(), false)
            .map(n -> n.get("option"))
            .filter(o -> {
                Optional<Long> id = SanitizingHelper.parse("id", o, Long.class);
                return id.isPresent() && persistedIds.contains(id.get());
            })
            .forEach(o -> updateOption(o, OptionUpdateOptions.SKIP_DEFAULTS));
        // Removals
        question.getOptions().stream().filter(o -> !providedIds.contains(o.getId())).forEach(this::deleteOption);
        // Additions
        StreamSupport
            .stream(node.spliterator(), false)
            .filter(o -> SanitizingHelper.parse("id", o, Long.class).isEmpty())
            .forEach(o -> createOptionBasedOnExamQuestion(question, esq, user, o));
        // Finally update own option scores:
        for (JsonNode option : node) {
            SanitizingHelper
                .parse("id", option, Long.class)
                .ifPresent(id -> {
                    ExamSectionQuestionOption esqo = DB.find(ExamSectionQuestionOption.class, id);
                    if (esqo != null) {
                        esqo.setScore(round(SanitizingHelper.parse("score", option, Double.class).orElse(null)));
                        esqo.update();
                    }
                });
        }
    }

    private boolean hasPositiveOptionScore(ArrayNode an) {
        return StreamSupport.stream(an.spliterator(), false).anyMatch(n -> n.get("score").asDouble() > 0);
    }

    private boolean hasValidClaimChoiceOptions(ArrayNode an) {
        boolean hasCorrectOption = StreamSupport
            .stream(an.spliterator(), false)
            .anyMatch(n -> {
                ClaimChoiceOptionType type = SanitizingHelper
                    .parseEnum("claimChoiceType", n.get("option"), ClaimChoiceOptionType.class)
                    .orElse(null);
                double score = n.get("score").asDouble();
                return type != ClaimChoiceOptionType.SkipOption && score > 0;
            });

        boolean hasIncorrectOption = StreamSupport
            .stream(an.spliterator(), false)
            .anyMatch(n -> {
                ClaimChoiceOptionType type = SanitizingHelper
                    .parseEnum("claimChoiceType", n.get("option"), ClaimChoiceOptionType.class)
                    .orElse(null);
                double score = n.get("score").asDouble();
                return type != ClaimChoiceOptionType.SkipOption && score <= 0;
            });

        boolean hasSkipOption = StreamSupport
            .stream(an.spliterator(), false)
            .anyMatch(n -> {
                ClaimChoiceOptionType type = SanitizingHelper
                    .parseEnum("claimChoiceType", n.get("option"), ClaimChoiceOptionType.class)
                    .orElse(null);
                double score = n.get("score").asDouble();
                return type == ClaimChoiceOptionType.SkipOption && score == 0;
            });

        return hasCorrectOption && hasIncorrectOption && hasSkipOption;
    }

    @Authenticated
    @With(SectionQuestionSanitizer.class)
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result updateDistributedExamQuestion(Long eid, Long sid, Long qid, Http.Request request) {
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        ExpressionList<ExamSectionQuestion> query = DB.find(ExamSectionQuestion.class).where().idEq(qid);
        if (user.hasRole(Role.Name.TEACHER)) {
            query = query.eq("examSection.exam.examOwners", user);
        }
        PathProperties pp = PathProperties.parse("(*, question(*, options(*)), options(*, option(*)))");
        query.apply(pp);
        ExamSectionQuestion examSectionQuestion = query.findOne();
        if (examSectionQuestion == null) {
            return forbidden("sitnet_error_access_forbidden");
        }
        Question question = DB
            .find(Question.class)
            .fetch("examSectionQuestions")
            .fetch("examSectionQuestions.options")
            .where()
            .idEq(examSectionQuestion.getQuestion().getId())
            .findOne();
        if (question == null) {
            return notFound();
        }
        JsonNode body = request.body().asJson();
        if (
            question.getType() == Question.Type.WeightedMultipleChoiceQuestion &&
            !hasPositiveOptionScore((ArrayNode) body.get("options"))
        ) {
            return badRequest("sitnet_correct_option_required");
        }

        if (
            question.getType() == Question.Type.ClaimChoiceQuestion &&
            !hasValidClaimChoiceOptions((ArrayNode) body.get("options"))
        ) {
            return badRequest("sitnet_incorrect_claim_question_options");
        }

        // Update question: text
        question.setQuestion(request.attrs().getOptional(Attrs.QUESTION_TEXT).orElse(null));
        question.update();
        updateExamQuestion(examSectionQuestion, body, request);
        examSectionQuestion.update();
        if (
            question.getType() != Question.Type.EssayQuestion && question.getType() != Question.Type.ClozeTestQuestion
        ) {
            // Process the options, this has an impact on the base question options as well as all the section questions
            // utilizing those.
            processExamQuestionOptions(question, examSectionQuestion, (ArrayNode) body.get("options"), user);
        }
        // Bit dumb, refetch from database to get the updated options right in response. Could be made more elegantly
        return ok(query.findOne(), pp);
    }

    @Authenticated
    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result updateUndistributedExamQuestion(Long eid, Long sid, Long qid, Http.Request request) {
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        ExpressionList<ExamSectionQuestion> query = DB.find(ExamSectionQuestion.class).where().idEq(qid);
        if (user.hasRole(Role.Name.TEACHER)) {
            query = query.eq("examSection.exam.examOwners", user);
        }
        PathProperties pp = PathProperties.parse("(*, question(*, attachment(*), options(*)), options(*, option(*)))");
        query.apply(pp);
        ExamSectionQuestion examSectionQuestion = query.findOne();
        if (examSectionQuestion == null) {
            return forbidden("sitnet_error_access_forbidden");
        }
        Question question = DB.find(Question.class, examSectionQuestion.getQuestion().getId());
        if (question == null) {
            return notFound();
        }
        updateExamQuestion(examSectionQuestion, question);
        examSectionQuestion.update();
        return ok(examSectionQuestion, pp);
    }

    @Restrict({ @Group("TEACHER"), @Group("ADMIN") })
    public Result getQuestionDistribution(Long id) {
        ExamSectionQuestion esq = DB.find(ExamSectionQuestion.class, id);
        if (esq == null) {
            return notFound();
        }
        Question question = esq.getQuestion();
        // ATM it is enough that question is bound to multiple exams
        boolean isDistributed =
            question.getExamSectionQuestions().stream().map(eq -> eq.getExamSection().getExam()).distinct().count() > 1;

        ObjectNode node = Json.newObject();
        node.put("distributed", isDistributed);
        return ok(Json.toJson(node));
    }
}

File: app/features/exam/controllers/ExamSectionController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package features.exam.controllers

import database.EbeanJsonExtensions
import features.exam.services.ExamSectionError.*
import features.exam.services.{ExamSectionError, ExamSectionService}
import io.ebean.text.PathProperties
import models.user.Role
import play.api.libs.json.{JsValue, Json}
import play.api.mvc.*
import security.Auth.{AuthenticatedAction, authorized}
import security.{Auth, BlockingIOExecutionContext}
import system.AuditedAction
import validation.core.PlayJsonHelper

import javax.inject.Inject
import scala.concurrent.Future

class ExamSectionController @Inject() (
    authenticated: AuthenticatedAction,
    audited: AuditedAction,
    private val examSectionService: ExamSectionService,
    val controllerComponents: ControllerComponents,
    implicit val ec: BlockingIOExecutionContext
) extends BaseController
    with EbeanJsonExtensions:

  private def toResult(error: ExamSectionError): Result =
    error match
      case ExamNotFound                  => NotFound("i18n_error_not_found")
      case SectionNotFound               => NotFound("i18n_error_not_found")
      case QuestionNotFound              => NotFound("i18n_error_not_found")
      case SectionQuestionNotFound       => NotFound("i18n_error_not_found")
      case AccessForbidden               => Forbidden("i18n_error_access_forbidden")
      case FutureReservationsExist       => Forbidden("i18n_error_future_reservations_exist")
      case AutoEvaluationEssayQuestion   => Forbidden("i18n_error_autoevaluation_essay_question")
      case QuestionAlreadyInSection      => BadRequest("i18n_question_already_in_section")
      case MissingSequenceNumber         => BadRequest("Missing sequenceNumber")
      case MissingQuestions              => BadRequest("Missing sequenceNumber or questions")
      case MissingFromOrTo               => BadRequest("Missing 'from' or 'to' parameter")
      case MissingOptionsArray           => BadRequest("Missing options array")
      case CorrectOptionRequired         => BadRequest("i18n_correct_option_required")
      case IncorrectClaimQuestionOptions => BadRequest("i18n_incorrect_claim_question_options")
      case ValidationError(message)      => BadRequest(message)

  def insertSection(eid: Long): Action[AnyContent] =
    audited.andThen(authenticated).andThen(authorized(Seq(
      Role.Name.TEACHER,
      Role.Name.ADMIN,
      Role.Name.SUPPORT
    ))) {
      request =>
        val user = request.attrs(Auth.ATTR_USER)
        examSectionService.insertSection(eid, user) match
          case Left(error) => toResult(error)
          case Right(section) =>
            Ok(section.asJson(PathProperties.parse("(*, examMaterials(*), sectionQuestions(*))")))
    }

  def removeSection(eid: Long, sid: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT))) {
      request =>
        val user = request.attrs(Auth.ATTR_USER)
        examSectionService.removeSection(eid, sid, user) match
          case Left(error) => toResult(error)
          case Right(_)    => Ok
    }

  def updateSection(eid: Long, sid: Long): Action[JsValue] =
    audited
      .andThen(authenticated)
      .andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT)))
      .async(parse.json) { request =>
        val user             = request.attrs(Auth.ATTR_USER)
        val body             = request.body
        val name             = PlayJsonHelper.parse[String]("name", body)
        val expanded         = PlayJsonHelper.parseOrElse("expanded", body, false)
        val lotteryOn        = PlayJsonHelper.parseOrElse("lotteryOn", body, false)
        val lotteryItemCount = PlayJsonHelper.parseOrElse("lotteryItemCount", body, 1)
        val description      = PlayJsonHelper.parse[String]("description", body)
        val optional         = PlayJsonHelper.parseOrElse("optional", body, false)

        examSectionService.updateSection(
          eid,
          sid,
          user,
          name,
          expanded,
          lotteryOn,
          lotteryItemCount,
          description,
          optional
        ) match
          case Left(error)    => Future.successful(toResult(error))
          case Right(section) => Future.successful(Ok(section.asJson))
      }

  def reorderSections(eid: Long): Action[JsValue] =
    audited
      .andThen(authenticated)
      .andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT)))
      .async(parse.json) { request =>
        val body    = request.body
        val fromOpt = PlayJsonHelper.parse[Int]("from", body)
        val toOpt   = PlayJsonHelper.parse[Int]("to", body)

        (fromOpt, toOpt) match
          case (Some(from), Some(to)) =>
            examSectionService.reorderSections(eid, from, to, request.attrs(Auth.ATTR_USER)) match
              case Left(error) => Future.successful(toResult(error))
              case Right(_)    => Future.successful(Ok)
          case _ => Future.successful(toResult(MissingFromOrTo))
      }

  def reorderSectionQuestions(eid: Long, sid: Long): Action[JsValue] =
    audited
      .andThen(authenticated)
      .andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT)))
      .async(parse.json) { request =>
        val body    = request.body
        val fromOpt = PlayJsonHelper.parse[Int]("from", body)
        val toOpt   = PlayJsonHelper.parse[Int]("to", body)

        (fromOpt, toOpt) match
          case (Some(from), Some(to)) =>
            examSectionService.reorderSectionQuestions(
              eid,
              sid,
              from,
              to,
              request.attrs(Auth.ATTR_USER)
            ) match
              case Left(error) => Future.successful(toResult(error))
              case Right(_)    => Future.successful(Ok)
          case _ => Future.successful(toResult(MissingFromOrTo))
      }

  def insertQuestion(eid: Long, sid: Long, qid: Long): Action[JsValue] =
    audited
      .andThen(authenticated)
      .andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT)))
      .async(parse.json) { request =>
        val user   = request.attrs(Auth.ATTR_USER)
        val body   = request.body
        val seqOpt = PlayJsonHelper.parse[Int]("sequenceNumber", body)

        seqOpt match
          case None => Future.successful(toResult(MissingSequenceNumber))
          case Some(seq) =>
            examSectionService.insertQuestion(eid, sid, qid, seq, user) match
              case Left(error)    => Future.successful(toResult(error))
              case Right(section) => Future.successful(Ok(section.asJson))
      }

  def insertMultipleQuestions(eid: Long, sid: Long): Action[JsValue] =
    audited
      .andThen(authenticated)
      .andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT)))
      .async(parse.json) { request =>
        val user         = request.attrs(Auth.ATTR_USER)
        val body         = request.body
        val sequenceOpt  = PlayJsonHelper.parse[Int]("sequenceNumber", body)
        val questionsOpt = PlayJsonHelper.parseCommaSeparatedLongs("questions", body)

        (sequenceOpt, questionsOpt) match
          case (Some(sequence), Some(questions)) =>
            examSectionService.insertMultipleQuestions(eid, sid, sequence, questions, user) match
              case Left(error)    => Future.successful(toResult(error))
              case Right(section) => Future.successful(Ok(section.asJson))
          case _ => Future.successful(toResult(MissingQuestions))
      }

  def removeQuestion(eid: Long, sid: Long, qid: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT))) {
      request =>
        val user = request.attrs(Auth.ATTR_USER)
        examSectionService.removeQuestion(eid, sid, qid, user) match
          case Left(error)    => toResult(error)
          case Right(section) => Ok(section.asJson)
    }

  def clearQuestions(eid: Long, sid: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT))) {
      request =>
        val user = request.attrs(Auth.ATTR_USER)
        examSectionService.clearQuestions(eid, sid, user) match
          case Left(error)    => toResult(error)
          case Right(section) => Ok(section.asJson)
    }

  def updateDistributedExamQuestion(eid: Long, sid: Long, qid: Long): Action[JsValue] =
    audited
      .andThen(authenticated)
      .andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT)))
      .async(parse.json) { request =>
        val user = request.attrs(Auth.ATTR_USER)
        val body = request.body

        examSectionService.updateDistributedExamQuestion(eid, sid, qid, user, body) match
          case Left(error) => Future.successful(toResult(error))
          case Right(examSectionQuestion) =>
            val pp = PathProperties.parse("(*, question(*, options(*)), options(*, option(*)))")
            Future.successful(Ok(examSectionQuestion.asJson(pp)))
      }

  def updateUndistributedExamQuestion(eid: Long, sid: Long, qid: Long): Action[AnyContent] =
    audited.andThen(authenticated).andThen(authorized(Seq(
      Role.Name.TEACHER,
      Role.Name.ADMIN,
      Role.Name.SUPPORT
    ))) {
      request =>
        val user = request.attrs(Auth.ATTR_USER)
        examSectionService.updateUndistributedExamQuestion(eid, sid, qid, user) match
          case Left(error) => toResult(error)
          case Right(examSectionQuestion) =>
            val pp = PathProperties.parse(
              "(*, question(*, attachment(*), options(*)), options(*, option(*)))"
            )
            Ok(examSectionQuestion.asJson(pp))
    }

  def getQuestionDistribution(id: Long): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT))) {
      _ =>
        examSectionService.getQuestionDistribution(id) match
          case Left(error)          => toResult(error)
          case Right(isDistributed) => Ok(Json.obj("distributed" -> isDistributed))
    }

  def listSections(
      filter: Option[String],
      courseIds: Option[List[Long]],
      examIds: Option[List[Long]],
      tagIds: Option[List[Long]],
      ownerIds: Option[List[Long]]
  ): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.TEACHER, Role.Name.ADMIN, Role.Name.SUPPORT))) {
      request =>
        val user = request.attrs(Auth.ATTR_USER)
        val sections =
          examSectionService.listSections(user, filter, courseIds, examIds, tagIds, ownerIds)
        Ok(sections.asJson(PathProperties.parse("(*, creator(id))")))
    }
